File: gtci/11-tree-breadth-first-search/python3/8-right-view-of-a-binary-tree.py
from collections import deque


class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None


class Solution:
    def traverse(self, root):
        if root is None:
            return []

        right_view = []

        queue = deque([root])

        while queue:
            level_size = len(queue)

            for i in range(level_size):
                node = queue.popleft()

                if i == level_size - 1:
                    right_view.append(node.val)

                if node.left is not None:
                    queue.append(node.left)

                if node.right is not None:
                    queue.append(node.right)

        return right_view


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.right.right = TreeNode(4)
    root.right.right.right = TreeNode(5)
    root.right.right.right.right = TreeNode(6)
    assert sol.traverse(root) == [1, 3, 4, 5, 6]

    root = TreeNode(12)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(9)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(5)
    assert sol.traverse(root) == [12, 1, 5]

    print("All test cases passed.")

File: gtci/11-tree-breadth-first-search/cpp/8-right-view-of-a-binary-tree.cpp
#include <cassert>
#include <iostream>
#include <vector>
#include <queue>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
  static vector<int> traverse(TreeNode* root) {
    if (root == nullptr)
      return {};

    vector<int> rightView;

    std::queue<TreeNode*> queue;
    queue.push(root);

    while (!queue.empty()) {
      const size_t levelSize = queue.size();

      for (size_t i = 0; i < levelSize; i++) {
        TreeNode* node = queue.front();
        queue.pop();

        if (i == levelSize - 1)
          rightView.push_back(node->val);
        
        if (node->left != nullptr)
          queue.push(node->left);
        
        if (node->right != nullptr)
          queue.push(node->right);
      }
    }

    return rightView;
  }
};

int main() {
    Solution sol = Solution();
    
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->right = new TreeNode(6);
    std::vector<int> expected1 = {1, 3, 6};
    assert(sol.traverse(root) == expected1);

    TreeNode* root2 = new TreeNode(12);
    root2->left = new TreeNode(7);
    root2->right = new TreeNode(1);
    root2->left->left = new TreeNode(9);
    root2->right->left = new TreeNode(10);
    root2->right->right = new TreeNode(5);
    std::vector<int> expected2 = {12, 1, 5};
    assert(sol.traverse(root2) == expected2);

    cout << "All test cases passed." << endl;

    delete root;
    delete root2;

    return 0;
}
